File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/provider/Query.java
/**
 * @file
 * @brief Playlist query helper
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.provider;

import android.content.ContentResolver;
import android.content.UriMatcher;
import android.net.Uri;

import androidx.annotation.IntDef;

import java.lang.annotation.Retention;
import java.util.List;

import static java.lang.annotation.RetentionPolicy.SOURCE;

/*
 * ${path} is full data uri (including subpath in fragment) stored as string
 *
 * content://app.zxtune.vfs/resolve/${path} - get object properties by full path
 * content://app.zxtune.vfs/listing/${path} - get directory content by full path
 * content://app.zxtune.vfs/parents/${path} - get object parents chain
 * content://app.zxtune.vfs/search/${path}?query=${query} - start search
 * content://app.zxtune.vfs/file/${path} - get information about and file object about local file
 */

class Query {

  @Retention(SOURCE)
  @IntDef({TYPE_RESOLVE, TYPE_LISTING, TYPE_PARENTS, TYPE_SEARCH, TYPE_FILE})
  @interface Type {}

  static final int TYPE_RESOLVE = 0;
  static final int TYPE_LISTING = 1;
  static final int TYPE_PARENTS = 2;
  static final int TYPE_SEARCH = 3;
  static final int TYPE_FILE = 4;

  private static final String AUTHORITY = "app.zxtune.vfs";
  private static final String RESOLVE_PATH = "resolve";
  private static final String LISTING_PATH = "listing";
  private static final String PARENTS_PATH = "parents";
  private static final String SEARCH_PATH = "search";
  private static final String QUERY_PARAM = "query";
  private static final String FILE_PATH = "file";

  private static final String ITEM_SUBTYPE = "vnd." + AUTHORITY + ".item";
  private static final String SIMPLE_ITEM_SUBTYPE = "vnd." + AUTHORITY + ".simple_item";

  private static final String MIME_ITEM =
      ContentResolver.CURSOR_ITEM_BASE_TYPE + "/" + ITEM_SUBTYPE;
  private static final String MIME_ITEMS_SET =
      ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + ITEM_SUBTYPE;
  private static final String MIME_SIMPLE_ITEMS_SET =
      ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + SIMPLE_ITEM_SUBTYPE;

  private static final UriMatcher uriTemplate;

  static {
    uriTemplate = new UriMatcher(UriMatcher.NO_MATCH);
    uriTemplate.addURI(AUTHORITY, RESOLVE_PATH, TYPE_RESOLVE);
    uriTemplate.addURI(AUTHORITY, RESOLVE_PATH + "/*", TYPE_RESOLVE);
    uriTemplate.addURI(AUTHORITY, LISTING_PATH, TYPE_LISTING);
    uriTemplate.addURI(AUTHORITY, LISTING_PATH + "/*", TYPE_LISTING);
    uriTemplate.addURI(AUTHORITY, PARENTS_PATH, TYPE_PARENTS);
    uriTemplate.addURI(AUTHORITY, PARENTS_PATH + "/*", TYPE_PARENTS);
    uriTemplate.addURI(AUTHORITY, SEARCH_PATH + "*", TYPE_SEARCH);
    uriTemplate.addURI(AUTHORITY, SEARCH_PATH + "/*", TYPE_SEARCH);
    uriTemplate.addURI(AUTHORITY, FILE_PATH + "/*", TYPE_FILE);
  }

  //! @return Mime type of uri (used in content provider)
  static String mimeTypeOf(Uri uri) {
    switch (uriTemplate.match(uri)) {
      case TYPE_RESOLVE:
      case TYPE_FILE:
        return MIME_ITEM;
      case TYPE_LISTING:
      case TYPE_SEARCH:
        return MIME_ITEMS_SET;
      case TYPE_PARENTS:
        return MIME_SIMPLE_ITEMS_SET;
      default:
        throw new IllegalArgumentException("Wrong URI: " + uri);
    }
  }

  static Uri getPathFrom(Uri uri) {
    switch (uriTemplate.match(uri)) {
      case TYPE_RESOLVE:
      case TYPE_LISTING:
      case TYPE_PARENTS:
      case TYPE_SEARCH:
      case TYPE_FILE:
        final List<String> segments = uri.getPathSegments();
        return segments.size() > 1 ? Uri.parse(segments.get(1)) : Uri.EMPTY;
      default:
        throw new IllegalArgumentException("Wrong URI: " + uri);
    }
  }

  static String getQueryFrom(Uri uri) {
    if (uriTemplate.match(uri) == TYPE_SEARCH) {
      final String query = uri.getQueryParameter(QUERY_PARAM);
      if (query != null) {
        return query;
      }
    }
    throw new IllegalArgumentException("Wrong search URI: " + uri);
  }

  @Type
  static int getUriType(Uri uri) {
    return uriTemplate.match(uri);
  }

  static Uri resolveUriFor(Uri uri) {
    return makeUri(RESOLVE_PATH, uri).build();
  }

  static Uri listingUriFor(Uri uri) {
    return makeUri(LISTING_PATH, uri).build();
  }

  static Uri parentsUriFor(Uri uri) {
    return makeUri(PARENTS_PATH, uri).build();
  }

  static Uri searchUriFor(Uri uri, String query) {
    return makeUri(SEARCH_PATH, uri).appendQueryParameter(QUERY_PARAM, query).build();
  }

  static Uri fileUriFor(Uri uri) {
    return makeUri(FILE_PATH, uri).build();
  }

  private static Uri.Builder makeUri(String path, Uri uri) {
    return new Uri.Builder()
               .scheme(ContentResolver.SCHEME_CONTENT)
               .authority(AUTHORITY)
               .encodedPath(path)
               .appendPath(uri.toString());
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/provider/Query.kt
/**
 * @file
 * @brief VFS provider query helper
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs.provider

import android.content.ContentResolver
import android.content.UriMatcher
import android.net.Uri
import androidx.core.net.toUri
import app.zxtune.BuildConfig

/*
 * ${path} is full data uri (including subpath in fragment) stored as string
 *
 * content://app.zxtune.vfs/resolve/${path} - get object properties by full path and all parents
 * content://app.zxtune.vfs/listing/${path} - get directory content by full path
 * content://app.zxtune.vfs/feed/${path} - VfsExtensions.FEED values
 * content://app.zxtune.vfs/search/${path}?query=${query} - start search
 * content://app.zxtune.vfs/file/${path}?size=${size} - get information/content of track file
 * content://app.zxtune.vfs/notification/${path} - get path-related notification
 */
internal class Query private constructor(val providerUri: Uri, val type: Type?) {
    enum class Type(val path: String, val mime: String) {
        RESOLVE("resolve", MIME_ITEMS_SET),

        LISTING("listing", MIME_ITEMS_SET),

        FEED("feed", MIME_ITEMS_SET),

        SEARCH("search", MIME_ITEMS_SET),

        FILE("file", MIME_ITEM),

        NOTIFICATION("notification", MIME_NOTIFICATION),
    }

    companion object {
        private const val AUTHORITY = "${BuildConfig.APPLICATION_ID}.vfs"
        private const val QUERY_PARAM = "query"
        private const val SIZE_PARAM = "size"
        private const val ITEM_SUBTYPE = "vnd.$AUTHORITY.item"
        private const val NOTIFICATION_SUBTYPE = "vnd.$AUTHORITY.notification"
        private const val MIME_ITEM = "${ContentResolver.CURSOR_ITEM_BASE_TYPE}/$ITEM_SUBTYPE"
        private const val MIME_ITEMS_SET = "${ContentResolver.CURSOR_DIR_BASE_TYPE}/$ITEM_SUBTYPE"
        private const val MIME_NOTIFICATION =
            "${ContentResolver.CURSOR_ITEM_BASE_TYPE}/$NOTIFICATION_SUBTYPE"
        private val uriTemplate = UriMatcher(UriMatcher.NO_MATCH).apply {
            Type.entries.forEach {
                // Empty path for empty root url
                addURI(AUTHORITY, it.path, it.ordinal)
                addURI(AUTHORITY, "${it.path}/*", it.ordinal)
            }
        }

        fun parse(uri: Uri) = Query(uri, Type.entries.getOrNull(uriTemplate.match(uri)))
        fun parse(uri: String) = parse(uri.toUri())

        fun forResolve(uri: Uri) = makeSimple(Type.RESOLVE, uri)

        fun forListing(uri: Uri) = makeSimple(Type.LISTING, uri)

        fun forFeed(uri: Uri) = makeSimple(Type.FEED, uri)

        fun forSearch(uri: Uri, query: String) =
            makeComplex(Type.SEARCH, uri) { appendQueryParameter(QUERY_PARAM, query) }

        fun forFile(uri: Uri, size: Long) =
            makeComplex(Type.FILE, uri) { appendQueryParameter(SIZE_PARAM, size.toString()) }

        fun forNotification(uri: Uri) = makeSimple(Type.NOTIFICATION, uri)

        private fun makeUri(type: Type, uri: Uri) =
            Uri.Builder().scheme(ContentResolver.SCHEME_CONTENT).authority(AUTHORITY)
                .encodedPath(type.path).appendPath(uri.toString())

        private fun makeSimple(type: Type, uri: Uri) = Query(makeUri(type, uri).build(), type)

        private fun makeComplex(type: Type, uri: Uri, patch: Uri.Builder.() -> Unit) =
            Query(makeUri(type, uri).apply(patch).build(), type)
    }

    val path: Uri
        get() = when (type) {
            Type.RESOLVE, Type.LISTING, Type.FEED, Type.SEARCH, Type.FILE, Type.NOTIFICATION -> providerUri.pathSegments.getOrNull(
                1
            )?.toUri() ?: Uri.EMPTY

            else -> throw IllegalArgumentException("Wrong URI: $providerUri")
        }

    val searchQuery
        get() = providerUri.takeIf { type == Type.SEARCH }?.getQueryParameter(QUERY_PARAM)
            ?: throw IllegalArgumentException("Wrong search URI: $providerUri")

    val fileSize
        get() = providerUri.takeIf { type == Type.FILE }?.getQueryParameter(SIZE_PARAM)
            ?.toLongOrNull() ?: throw IllegalArgumentException("Wrong file URI: $providerUri")
}
